From c6dcfccdd6b3bf1898d8056d909b962588580570 Mon Sep 17 00:00:00 2001
From: Wenzong Fan <wenzong.fan@windriver.com>
Date: Mon, 27 Feb 2017 10:33:18 +0000
Subject: [PATCH] refpolicy: fix avc denials

Reference sources: selinux-policy-3.13.1-225.6.fc25.src.rpm
Changes come from: policy-f25-base.patch
                   policy-f25-contrib.patch

* allow NetworkManager to read:
  /run/systemd/seats (systemd_logind_var_run_t)
  /run/systemd/machines (init_var_run_t)
  /run/systemd/sessions (systemd_logind_sessions_t)

* allow NetworkManager to write /sys (sysfs_t)
* allow acpid (apmd_t) to search /run/systemd/journal (syslogd_var_run_t)
* allow dhclient to search /etc/ssl (cert_t), /run/systemd/journal
* allow named to search /proc/net (sysctl_net_t)
* allow polkitd to read:
  /run/systemd/seats
  /run/systemd/sessions
  /proc/sys/vm/overcommit_memory (sysctl_vm_overcommit_t)
  /run/systemd/machines

* allow polkitd to search:
  /run/systemd/journal
  /proc/sys/vm (sysctl_vm_t)

* allow systemd-hostnamed to read sysfs_t file/lnk_file, search /run/systemd/journal and tmpfs dir
* allow systemd-logind to create, mounton /var/run/user/* (user_runtime_t)
* allow systemd-logind to start/stop/status init_var_run_t services:
  /run/systemd/transient/user-0.slice
  /run/systemd/transient/session-c1.scope

* allow systemd-user-sessions to write /dev/kmsg (kmsg_device_t)
* allow X to read and write /dev/mem (memory_device_t)

WRL Fixes:
* allow irqbalance to write /proc/irq/* dirs (sysctl_irq_t), create
  smp_affinity files (sysctl_irq_t)
* allow zebra/ripngd/ospf6d/ospfd/ripd/bgpd (zebra_t) to read nsfs_t files
* dontaudit systemd-user-sessions read file_contexts.subs_dist (file_context_t)

Fix avc denials:
  avc: denied { read } for pid=452 comm="NetworkManager" \
  name="seats" dev="tmpfs" ino=1494 \
  scontext=system_u:system_r:NetworkManager_t:s0-s15:c0.c1023 \
  tcontext=system_u:object_r:systemd_logind_var_run_t:s0 tclass=dir

  avc: denied { read } for pid=466 comm="NetworkManager" \
  name="sessions" dev="tmpfs" ino=12334 \
  scontext=system_u:system_r:NetworkManager_t:s0-s15:c0.c1023 \
  tcontext=system_u:object_r:systemd_logind_sessions_t:s0 tclass=dir

  avc: denied { read } for pid=449 comm="NetworkManager" \
  name="machines" dev="tmpfs" ino=1667 \
  scontext=system_u:system_r:NetworkManager_t:s0-s15:c0.c1023 \
  tcontext=system_u:object_r:init_var_run_t:s0 tclass=dir

  avc: denied { write } for pid=452 comm="NetworkManager" \
  name="/" dev="sysfs" ino=1 \
  scontext=system_u:system_r:NetworkManager_t:s0-s15:c0.c1023 \
  tcontext=system_u:object_r:sysfs_t:s0 tclass=dir

  avc: denied { search } for pid=473 comm="acpid" \
  name="journal" dev="tmpfs" ino=12118 \
  scontext=system_u:system_r:apmd_t:s0-s15:c0.c1023 \
  tcontext=system_u:object_r:syslogd_var_run_t:s15:c0.c1023 tclass=dir

  avc: denied { search } for pid=983 comm="dhclient" \
  name="ssl" dev="sda1" ino=1358436 \
  scontext=system_u:system_r:dhcpc_t:s0-s15:c0.c1023 \
  tcontext=system_u:object_r:cert_t:s0 tclass=dir

  avc: denied { search } for pid=983 comm="dhclient" \
  name="journal" dev="tmpfs" ino=12118 \
  scontext=system_u:system_r:dhcpc_t:s0-s15:c0.c1023 \
  tcontext=system_u:object_r:syslogd_var_run_t:s15:c0.c1023 tclass=dir

  avc: denied { write } for pid=440 comm="irqbalance" \
  name="40" dev="proc" ino=4026531956 \
  scontext=system_u:system_r:irqbalance_t:s0-s15:c0.c1023 \
  tcontext=system_u:object_r:sysctl_irq_t:s0 tclass=dir

  avc: denied { create } for pid=420 comm="irqbalance" \
  name="smp_affinity" \
  scontext=system_u:system_r:irqbalance_t:s0-s15:c0.c1023 \
  tcontext=system_u:object_r:sysctl_irq_t:s0 tclass=file

  avc: denied { associate } for pid=423 comm="irqbalance" \
  name="smp_affinity" \
  scontext=system_u:object_r:sysctl_irq_t:s0 \
  tcontext=system_u:object_r:proc_t:s0 tclass=filesystem

  avc: denied { search } for pid=791 comm="named" \
  name="net" dev="proc" ino=2658 \
  scontext=system_u:system_r:named_t:s0-s15:c0.c1023 \
  tcontext=system_u:object_r:sysctl_net_t:s0 tclass=dir

  avc: denied { search } for pid=818 comm="polkitd" \
  name="journal" dev="tmpfs" ino=12118 \
  scontext=system_u:system_r:policykit_t:s0-s15:c0.c1023 \
  tcontext=system_u:object_r:syslogd_var_run_t:s15:c0.c1023 tclass=dir

  avc: denied { read } for pid=818 comm="polkitd" \
  name="seats" dev="tmpfs" ino=1494 \
  scontext=system_u:system_r:policykit_t:s0-s15:c0.c1023 \
  tcontext=system_u:object_r:systemd_logind_var_run_t:s0 tclass=dir

  avc: denied { search } for pid=841 \
  comm=4A5320536F75727E20546872656164 name="vm" dev="proc" ino=15932 \
  scontext=system_u:system_r:policykit_t:s0-s15:c0.c1023 \
  tcontext=system_u:object_r:sysctl_vm_t:s0 tclass=dir

  avc: denied { read } for pid=808 comm="polkitd" \
  name="sessions" dev="tmpfs" ino=12334 \
  scontext=system_u:system_r:policykit_t:s0-s15:c0.c1023 \
  tcontext=system_u:object_r:systemd_logind_sessions_t:s0 tclass=dir

  avc: denied { read } for pid=846 comm=4A5320536F75727E20546872656164 \
  name="overcommit_memory" dev="proc" ino=17506 \
  scontext=system_u:system_r:policykit_t:s0-s15:c0.c1023 \
  tcontext=system_u:object_r:sysctl_vm_overcommit_t:s0 tclass=file

  avc: denied { read } for pid=816 comm="polkitd" \
  name="machines" dev="tmpfs" ino=1667 \
  scontext=system_u:system_r:policykit_t:s0-s15:c0.c1023 \
  tcontext=system_u:object_r:init_var_run_t:s0 tclass=dir

  avc: denied { search } for pid=505 comm="systemd-hostnam" \
  name="journal" dev="tmpfs" ino=12118 \
  scontext=system_u:system_r:systemd_hostnamed_t:s0-s15:c0.c1023 \
  tcontext=system_u:object_r:syslogd_var_run_t:s15:c0.c1023 tclass=dir

  avc: denied { search } for pid=505 comm="systemd-hostnam" \
  name="/" dev="tmpfs" ino=17436 \
  scontext=system_u:system_r:systemd_hostnamed_t:s0-s15:c0.c1023 \
  tcontext=system_u:object_r:tmpfs_t:s0 tclass=dir

  avc: denied { read } for pid=505 comm="systemd-hostnam" \
  name="id" dev="sysfs" ino=220 \
  scontext=system_u:system_r:systemd_hostnamed_t:s0-s15:c0.c1023 \
  tcontext=system_u:object_r:sysfs_t:s0 tclass=lnk_file

  avc: denied { read } for pid=505 comm="systemd-hostnam" \
  name="pm_profile" dev="sysfs" ino=1390 \
  scontext=system_u:system_r:systemd_hostnamed_t:s0-s15:c0.c1023 \
  tcontext=system_u:object_r:sysfs_t:s0 tclass=file

  avc: denied { create } for pid=414 comm="systemd-logind" name="0" \
  scontext=system_u:system_r:systemd_logind_t:s0-s15:c0.c1023 \
  tcontext=root:object_r:user_runtime_t:s0 tclass=dir

  avc: denied { start } for auid=n/a uid=0 gid=0 \
  path="/run/systemd/transient/user-0.slice" \
  cmdline="/lib/systemd/systemd-logind" \
  scontext=system_u:system_r:systemd_logind_t:s0-s15:c0.c1023 \
  tcontext=system_u:object_r:init_var_run_t:s0 tclass=service

  avc: denied { status } for auid=n/a uid=0 gid=0 \
  path="/run/systemd/transient/session-c1.scope" \
  cmdline="/lib/systemd/systemd-logind" \
  scontext=system_u:system_r:systemd_logind_t:s0-s15:c0.c1023 \
  tcontext=system_u:object_r:init_var_run_t:s0 tclass=service

  avc: denied { write } for pid=557 comm="systemd-user-se" \
  name="kmsg" dev="devtmpfs" ino=2061 \
  scontext=system_u:system_r:systemd_sessions_t:s0-s15:c0.c1023 \
  tcontext=system_u:object_r:kmsg_device_t:s15:c0.c1023 tclass=chr_file

  avc: denied { read write } for pid=763 comm="X" \
  name="mem" dev="devtmpfs" ino=8400 \
  scontext=system_u:system_r:xserver_t:s0-s15:c0.c1023 \
  tcontext=system_u:object_r:memory_device_t:s15:c0.c1023 tclass=chr_file

  avc: denied { read } for pid=487 comm="zebra" \
  dev="nsfs" ino=4026531969 \
  scontext=system_u:system_r:zebra_t:s0-s15:c0.c1023 \
  tcontext=system_u:object_r:nsfs_t:s0 tclass=file

  avc: denied { read } for pid=684 comm="systemd-user-se" \
  name="file_contexts.subs_dist" dev="sda1" ino=1766526 \
  scontext=system_u:system_r:systemd_sessions_t:s0-s15:c0.c1023 \
  tcontext=system_u:object_r:file_context_t:s0 tclass=file

Upstream-Status: Pending

Signed-off-by: Wenzong Fan <wenzong.fan@windriver.com>
---
 policy/modules/services/acpi.te           |  1 +
 policy/modules/services/irqbalance.te     |  6 ++++++
 policy/modules/services/networkmanager.te |  4 ++++
 policy/modules/services/policykit.te      |  6 ++++++
 policy/modules/services/xserver.te        |  3 +++
 policy/modules/services/zebra.te          |  1 +
 policy/modules/system/init.if             | 19 +++++++++++++++++++
 policy/modules/system/systemd.if          | 19 +++++++++++++++++++
 policy/modules/system/systemd.te          | 14 +++++++++++++-
 9 files changed, 72 insertions(+), 1 deletion(-)

diff --git a/policy/modules/services/acpi.te b/policy/modules/services/acpi.te
index 3a7320d..5a7741d 100644
--- a/policy/modules/services/acpi.te
+++ b/policy/modules/services/acpi.te
@@ -98,6 +98,7 @@ files_var_lib_filetrans(acpid_t, acpid_var_lib_t, dir)
 manage_files_pattern(acpid_t, acpid_var_run_t, acpid_var_run_t)
 manage_sock_files_pattern(acpid_t, acpid_var_run_t, acpid_var_run_t)
 files_pid_filetrans(acpid_t, acpid_var_run_t, { file sock_file })
+mls_file_read_all_levels(acpid_t)
 
 can_exec(acpid_t, acpid_var_run_t)
 
diff --git a/policy/modules/services/irqbalance.te b/policy/modules/services/irqbalance.te
index ffc5997..1f0ce10 100644
--- a/policy/modules/services/irqbalance.te
+++ b/policy/modules/services/irqbalance.te
@@ -38,6 +38,12 @@ kernel_read_network_state(irqbalance_t)
 kernel_read_system_state(irqbalance_t)
 kernel_read_kernel_sysctls(irqbalance_t)
 kernel_rw_irq_sysctls(irqbalance_t)
+gen_require(`
+	type sysctl_irq_t, proc_t;
+')
+allow irqbalance_t sysctl_irq_t:dir rw_dir_perms;
+allow irqbalance_t sysctl_irq_t:file create;
+allow sysctl_irq_t proc_t:filesystem associate;
 
 dev_read_sysfs(irqbalance_t)
 
diff --git a/policy/modules/services/networkmanager.te b/policy/modules/services/networkmanager.te
index aaa0a8a..7c9f4c3 100644
--- a/policy/modules/services/networkmanager.te
+++ b/policy/modules/services/networkmanager.te
@@ -171,6 +171,7 @@ storage_getattr_fixed_disk_dev(NetworkManager_t)
 init_read_utmp(NetworkManager_t)
 init_dontaudit_write_utmp(NetworkManager_t)
 init_domtrans_script(NetworkManager_t)
+init_create_pid_dirs(NetworkManager_t)
 
 auth_use_nsswitch(NetworkManager_t)
 
@@ -196,6 +197,9 @@ sysnet_search_dhcp_state(NetworkManager_t)
 sysnet_manage_config(NetworkManager_t)
 sysnet_etc_filetrans_config(NetworkManager_t)
 
+systemd_login_list_pid_dirs(NetworkManager_t)
+systemd_read_logind_sessions_files(NetworkManager_t)
+
 # certificates in user home directories (cert_home_t in ~/\.pki)
 userdom_read_user_certs(NetworkManager_t)
 
diff --git a/policy/modules/services/policykit.te b/policy/modules/services/policykit.te
index 80cc9d8..96906a0 100644
--- a/policy/modules/services/policykit.te
+++ b/policy/modules/services/policykit.te
@@ -88,6 +88,8 @@ domtrans_pattern(policykit_t, policykit_resolve_exec_t, policykit_resolve_t)
 kernel_read_crypto_sysctls(policykit_t)
 kernel_read_kernel_sysctls(policykit_t)
 kernel_read_system_state(policykit_t)
+kernel_search_vm_sysctl(policykit_t)
+kernel_read_vm_overcommit_sysctl(policykit_t)
 
 dev_read_urand(policykit_t)
 dev_read_urand(policykit_t)
@@ -95,6 +97,10 @@ dev_read_urand(policykit_t)
 domain_read_all_domains_state(policykit_t)
 
 files_dontaudit_search_all_mountpoints(policykit_t)
+mls_file_read_all_levels(policykit_t)
+systemd_login_list_pid_dirs(policykit_t)
+systemd_read_logind_sessions_files(policykit_t)
+init_create_pid_dirs(policykit_t)
 
 fs_getattr_xattr_fs(policykit_t)
 fs_list_inotifyfs(policykit_t)
diff --git a/policy/modules/services/xserver.te b/policy/modules/services/xserver.te
index a2b08a8..2e7beb4 100644
--- a/policy/modules/services/xserver.te
+++ b/policy/modules/services/xserver.te
@@ -774,6 +774,9 @@ files_read_etc_files(xserver_t)
 files_read_etc_runtime_files(xserver_t)
 files_read_usr_files(xserver_t)
 
+mls_file_read_all_levels(xserver_t)
+mls_file_write_all_levels(xserver_t)
+
 # brought on by rhgb
 files_search_mnt(xserver_t)
 # for nscd
diff --git a/policy/modules/services/zebra.te b/policy/modules/services/zebra.te
index fe5a4fb..c34cca9 100644
--- a/policy/modules/services/zebra.te
+++ b/policy/modules/services/zebra.te
@@ -110,6 +110,7 @@ files_read_etc_runtime_files(zebra_t)
 
 fs_getattr_all_fs(zebra_t)
 fs_search_auto_mountpoints(zebra_t)
+fs_read_nsfs_files(zebra_t)
 
 term_list_ptys(zebra_t)
 
diff --git a/policy/modules/system/init.if b/policy/modules/system/init.if
index 7f73e0e..57affe4 100644
--- a/policy/modules/system/init.if
+++ b/policy/modules/system/init.if
@@ -2802,6 +2802,25 @@ interface(`init_rename_pid_files',`
 
 ########################################
 ## <summary>
+##      Allow the specified domain to modify the systemd configuration of
+##      transient scripts.
+## </summary>
+## <param name="domain">
+##      <summary>
+##      Domain allowed access.
+##      </summary>
+## </param>
+#
+interface(`init_config_transient_files',`
+	gen_require(`
+		type init_var_run_t;
+	')
+
+	allow $1 init_var_run_t:service { start stop status };
+')
+
+########################################
+## <summary>
 ##      Rename init_var_run_t files
 ## </summary>
 ## <param name="domain">
diff --git a/policy/modules/system/systemd.if b/policy/modules/system/systemd.if
index d381d44..871578f 100644
--- a/policy/modules/system/systemd.if
+++ b/policy/modules/system/systemd.if
@@ -116,6 +116,25 @@ interface(`systemd_write_logind_pid_pipes',`
 
 ######################################
 ## <summary>
+##      Read systemd_login PID files.
+## </summary>
+## <param name="domain">
+##      <summary>
+##      Domain allowed access.
+##      </summary>
+## </param>
+#
+interface(`systemd_login_list_pid_dirs',`
+	gen_require(`
+		type systemd_logind_var_run_t;
+	')
+
+	files_search_pids($1)
+	list_dirs_pattern($1, systemd_logind_var_run_t, systemd_logind_var_run_t)
+')
+
+######################################
+## <summary>
 ##   Use inherited systemd
 ##   logind file descriptors.
 ## </summary>
diff --git a/policy/modules/system/systemd.te b/policy/modules/system/systemd.te
index b3ffcfa..a901099 100644
--- a/policy/modules/system/systemd.te
+++ b/policy/modules/system/systemd.te
@@ -214,7 +214,7 @@ init_unit_file(power_unit_t)
 # Backlight local policy
 #
 
-allow systemd_backlight_t self:unix_dgram_socket { connect connected_socket_perms };
+allow systemd_backlight_t self:unix_dgram_socket { create connect connected_socket_perms };
 
 allow systemd_backlight_t systemd_backlight_var_lib_t:dir manage_dir_perms;
 init_var_lib_filetrans(systemd_backlight_t, systemd_backlight_var_lib_t, dir)
@@ -233,6 +233,10 @@ udev_read_pid_files(systemd_backlight_t)
 
 files_search_var_lib(systemd_backlight_t)
 
+dev_write_kmsg(systemd_backlight_t)
+mls_file_write_all_levels(systemd_backlight_t)
+kernel_read_system_state(systemd_backlight_t)
+
 #######################################
 #
 # Binfmt local policy
@@ -350,6 +354,9 @@ dev_read_sysfs(systemd_hostnamed_t)
 
 files_read_etc_files(systemd_hostnamed_t)
 
+fs_search_tmpfs(systemd_hostnamed_t)
+mls_file_read_all_levels(systemd_hostnamed_t)
+
 seutil_read_file_contexts(systemd_hostnamed_t)
 
 sysnet_etc_filetrans_config(systemd_hostnamed_t)
@@ -507,6 +514,8 @@ init_stop_system(systemd_logind_t)
 init_reboot_system(systemd_logind_t)
 init_shutdown_system(systemd_logind_t)
 
+init_config_transient_files(systemd_logind_t)
+
 init_dgram_send(systemd_logind_t)
 init_write_pid_socket(systemd_logind_t)
 kernel_read_system_state(systemd_logind_t)
@@ -519,6 +528,7 @@ domain_read_all_domains_state(systemd_logind_t)
 
 userdom_manage_all_user_tmp_content(systemd_logind_t)
 userdom_mounton_tmp_dirs(systemd_logind_t)
+userdom_manage_user_tmp_dirs(systemd_logind_t)
 
 files_dontaudit_list_non_security(systemd_logind_t)
 auth_dontaudit_mounton_var_auth(systemd_logind_t)
@@ -1016,6 +1026,7 @@ allow systemd_sessions_t self:process setfscreate;
 
 allow systemd_sessions_t systemd_sessions_var_run_t:file manage_file_perms;
 files_pid_filetrans(systemd_sessions_t, systemd_sessions_var_run_t, file)
+mls_file_write_all_levels(systemd_sessions_t)
 
 selinux_get_enforce_mode(systemd_sessions_t)
 selinux_get_fs_mount(systemd_sessions_t)
@@ -1025,6 +1036,7 @@ seutil_read_default_contexts(systemd_sessions_t)
 seutil_read_file_contexts(systemd_sessions_t)
 
 systemd_log_parse_environment(systemd_sessions_t)
+seutil_dontaudit_read_file_contexts(systemd_sessions_t)
 
 #########################################
 #
-- 
2.7.4

